### PR TITLE
Implement ht find

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Solves #61 

Note: I assume `s` from the paper being the size (number of buckets) of the hashtable.

### CODE REVIEW

File: src/libreset/ht.c [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+struct ht_bucket*
+ht_find(
+    struct ht* ht,
+    rs_hash hash
+) {
+    if (!ht) {
+        return NULL;
+    }
+
+    return &ht->buckets[hash % (CONSTPOW_TWO(BITCOUNT(hash)) / ht->nbuckets)];
```
Comment (neithernut) [inline src/libreset/ht.c]: Have a look at the paper. The bucket is _not_ selected by applying the modulo, but by "selecting" the most significant bits of the hash.

Comment (matthiasbeyer) [inline src/libreset/ht.c]: I already read this part of the paper, but I guess I do not understand the approach. Your patches are welcome!

Comment (matthiasbeyer) [inline src/libreset/ht.c]: Actually, this line is a implementation of fig 6 from the paper, so I do not know what's wrong here!

Comment (neithernut) [inline src/libreset/ht.c]: The equation goes `h(i) / (2^N/s)`, not `h(i) % (2^N/s)`! That's a division, not a modulo.

Comment (matthiasbeyer) [inline src/libreset/ht.c]: Actually not. The paper says modulo.

Comment (matthiasbeyer) [inline src/libreset/ht.c]: OMFG. This was a reading error. The character for division is the same as for modulo, in some way.

We should definitely fix this in the paper and use a colon there!

**Comment by matthiasbeyer** [comment]
@neithernut Can you tell what this error means (from the build):

```
/home/travis/build/matthiasbeyer/libreset/src/libreset/ht.c: In function ‘ht_find’:
/home/travis/build/matthiasbeyer/libreset/src/libreset/ht.c:41:5: error: left shift count >= width of type [-Werror]
     return &ht->buckets[hash / ((CONSTPOW_TWO(BITCOUNT(hash)) / ht->nbuckets))];
```

and how to fix it?

---

By the way: You wrote in the paper that we use the most significant **bits**, but you did not state how many bits...

**Comment by neithernut** [comment]
The error means that you shift so many bits to the left that the resulting value will always be `0`.

---

You must only use so many bits as there are neccessary to hold the buckets' indices. The term given in the paper should yield just that.

**Comment by matthiasbeyer** [comment]
Then, I'm not capable of this, as the implemented function literally _is_ the term from the paper.

**Comment by neithernut** [comment]
However, you run into an issue regarding the number of bits you have here. Hence, you have to calculate the value a bit differently.

**Comment by matthiasbeyer** [comment]
"Fixing" the macros with casts to `size_t` doesn't help.

**Comment by neithernut** [comment]
Yep, the forula has to be adjusted instead.
Since the number of buckets is always `2^x`, you may just store `x` in the ht data type. You can then just do something like:

``` c
bucket[hash >> (width_of_hash - x)]
```

**Comment by neithernut** [comment]
You could also calculate `x` on the fly, but why not just put it into the ht.

**Comment by neithernut** [comment]
Btw: this issue is a perfect example of how CI and -Werror help to remove bugs before they are introduced into the master/dev branch!

**Comment by matthiasbeyer** [comment]
Putting it into the `struct ht` type and replacing the `nbuckets` would cause more changes: The `_init` and `_destroy` function make use of the `nbuckets` member. This can be changed, of course, but then we would have to calculate the number of buckets there...

Of course, this would be much more efficient, as `_init` and `_destroy` are called just once for a `struct ht` object.

I will try to implement it this way.

**Comment by neithernut** [comment]
The information density is much higher if you just store the exponent.
Also, you can get `nbuckets` by doing a simple left-shift. Doing the opposite takes a few lines.

File: src/libreset/ht.h [hunk 2]
```diff
@@ -46,10 +46,12 @@ struct ht_bucket {
...
 /**
  * Hashtable type
+ *
+ * The type for the hashtable, holding buckets and size.
  */
 struct ht {
-    struct ht_bucket* buckets;
-    size_t nbuckets;
+    struct ht_bucket* buckets; //!< The buckets of the hashtable
+    size_t sizeexp; //!< Exponent 2 must be raised to, to get the size of the ht
```
Comment (neithernut) [inline src/libreset/ht.h]: please add a comma or something after "Exponent"

File: src/libreset/ht.c [hunk 3]
```diff
@@ -20,10 +21,24 @@ ht_destroy(
     struct ht* ht
 ) {
     if (ht) {
-        for (; ht->nbuckets > 0; ht->nbuckets--) {
-            avl_destroy(&ht->buckets[ht->nbuckets - 1].avl);
+        size_t i = ht_nbuckets(ht);
+        for (; i > 0; i--) {
+            avl_destroy(&ht->buckets[i - 1].avl);
```
Comment (neithernut) [inline src/libreset/ht.c]: Declaring an iterator before the loop and then having a 0-check as the condition seem odd.
Imho the following would be more elegant:

``` c
size_t i = ht_nbuckets(ht);
while (--i) {
    avl_destroy(&ht->buckets[i].avl);
}
```

File: src/libreset/ht.c [hunk 4]
```diff
@@ -20,10 +21,24 @@ ht_destroy(
     struct ht* ht
 ) {
     if (ht) {
```
Comment (neithernut) [inline src/libreset/ht.c:23]: Early return? Would be more elegant imho.

Comment (matthiasbeyer) [inline src/libreset/ht.c:23]: Nah, I don't care about it if there are less then 10 lines in the if branch.

``` c
if (foo) {
    bar();
    baz();
}
```

or

``` c
if (!foo) {
    return;
}
bar();
baz();
```

doesn't make that much difference.

File: src/libreset/ht.c [hunk 5]
```diff
@@ -20,10 +21,24 @@ ht_destroy(
...
-        for (; ht->nbuckets > 0; ht->nbuckets--) {
-            avl_destroy(&ht->buckets[ht->nbuckets - 1].avl);
+        size_t i = ht_nbuckets(ht);
+        while (--i) {
```
Comment (neithernut) [inline src/libreset/ht.c]: I made a mistake earlier. It should be `i--` rather than `--i`.

**Comment by neithernut** [comment]
ACK for squashing and merge.

**Pull request merged**